libraries/AP_Proximity/AP_Proximity_LD06.cpp

```diff
@@ -0,0 +1,183 @@
+        if(c == LD_START_CHAR || _response_data) {
+
+            // Sets to true if a start byte has been read, default false otherwise
+            _response_data = true;
+
+            // Stores the next byte in an array
+            _response[_byte_count] = c;
+            _byte_count++;
+
+            // The response should have 47 bytes in total
+            if (_byte_count == sizeof(_response)) {
+                parse_response_data();
+                _temp_boundary.update_3D_boundary(boundary);
+                _byte_count = 0;
+                _response_data = false;
+
+                set_status(AP_Proximity::Status::Good);
+            }
+        } else {
+            gcs().send_text(MAV_SEVERITY_CRITICAL, "LD06 Invalid Packet: %d", c);
+        }
+    }
+
+    // Store for debugging purposes
+    _last_request_ms = AP_HAL::millis();
+}
+
+// Parses the data packet received from the LiDAR
+void AP_Proximity_LD06::parse_response_data()
+{ 
+
+    // Data interpretation based on:
+    // http://wiki.inno-maker.com/display/HOMEPAGE/LD06?preview=/6949506/6949511/LDROBOT_LD06_Development%20manual_v1.0_en.pdf
+
+    // Second byte in array stores length of data - not used but stored for debugging
+    // const uint8_t data_length = _response[_start_data_length];
+
+    // Respective bits store the radar speed, start/end angles
+    // Use bitwise operations to correctly obtain correct angles
+    // Divide angles by 100 as per manual
+    // const uint16_t radar_speed = UINT16_VALUE(_response[_start_radar_speed + 1], _response[_start_radar_speed]);
+    const float start_angle = float(UINT16_VALUE(_response[START_BEGIN_ANGLE + 1], _response[START_BEGIN_ANGLE]))  * 0.01;
+    const float end_angle = float(UINT16_VALUE(_response[START_END_ANGLE + 1], _response[START_END_ANGLE])) * 0.01;
+
+    // Verify the checksum that is stored in the last element of the response array
+    // Return if checksum is incorrect - i.e. bad data, bad readings, etc.
+    uint8_t check_sum = _response[START_CHECK_SUM];
+    if(check_sum != calc_checksum(&_response[0], sizeof(_response)/sizeof(_response[0]) - 1)) {
+        gcs().send_text(MAV_SEVERITY_CRITICAL, "LD06 Bad Check Sum: %d", check_sum);
+        return;
+    }
+
+    // Used to calculate the angle inside the data packet
+    uint8_t angle_inc_counter = 0;
+
+    // Each recording point is three bytes long, goes through all of that and updates database
+    for (uint16_t i = START_PAYLOAD; i < START_PAYLOAD + MEASUREMENT_PAYLOAD_LENGTH * PAYLOAD_COUNT; i += MEASUREMENT_PAYLOAD_LENGTH) {
+
+        // Gets the distance recorded and converts to meters
+        const uint16_t distance_mm = (_response[i + 1] << 0x08) | (_response[i]);
+        float distance = distance_mm * 0.001f;
+
+        // Calculates the angle that this point was sampled at
+        const float sampled_angle = start_angle + angle_inc_counter * (end_angle - start_angle) /  (PAYLOAD_COUNT - 1);
```

Comment:
We can simplify this a bit and move some computation outside the loop.
```suggestion
        sampled_angle += angle_step;
```